From 662fd7393a15788ff3af5cad65c26a52ed8d15e8 Mon Sep 17 00:00:00 2001
From: Jay Petacat <jay@jayschwa.net>
Date: Fri, 31 Oct 2025 12:31:57 -0600
Subject: [PATCH 040/133] std: Skip element comparisons if `mem.order` args
 point to same memory
X-Developer-Signature: v=1; a=openpgp-sha256; l=3485; i=xtex@astrafall.org;
 h=from:subject; bh=sRQNyGvMsCaRL2JqFB0zguVaVcvQVuwR+rJQAUiMRTs=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn7cv+p/uKpvBuy+qO457WPNSc+7Rf1vD/hYHPPnQu72
 Z88Lt6c0zGRhUGMi8FSTJGlyLDBm1UnnV90WbkszBxWJpAh0iINDAwMDCwMfLmJeaVGOkZ6ptqG
 eoaGOkY6RgxcnAIw1VFTGf4Zakw7UMmQ6R5TILTYMKtmdneXkbfN1pqFEfI3IsI7d01m+GdwJV2
 9ZlMC787uywtPpXdUrje2WWistJhB5uLV7eXNRowA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

This optimization is used in `mem.eql`, but was missing from `order`,
`orderZ`, and `ascii.orderIgnoreCase`.

(cherry picked from commit 5a38dd28dc6eb7fe1acb7bb94710ab1a99a13565)
---
 lib/std/ascii.zig | 16 +++++++++-------
 lib/std/mem.zig   | 22 ++++++++++++++++------
 2 files changed, 25 insertions(+), 13 deletions(-)

diff --git a/lib/std/ascii.zig b/lib/std/ascii.zig
index 79f6d8a296ee..a70e33fdb938 100644
--- a/lib/std/ascii.zig
+++ b/lib/std/ascii.zig
@@ -420,13 +420,15 @@ test indexOfIgnoreCase {
 
 /// Returns the lexicographical order of two slices. O(n).
 pub fn orderIgnoreCase(lhs: []const u8, rhs: []const u8) std.math.Order {
-    const n = @min(lhs.len, rhs.len);
-    var i: usize = 0;
-    while (i < n) : (i += 1) {
-        switch (std.math.order(toLower(lhs[i]), toLower(rhs[i]))) {
-            .eq => continue,
-            .lt => return .lt,
-            .gt => return .gt,
+    if (lhs.ptr != rhs.ptr) {
+        const n = @min(lhs.len, rhs.len);
+        var i: usize = 0;
+        while (i < n) : (i += 1) {
+            switch (std.math.order(toLower(lhs[i]), toLower(rhs[i]))) {
+                .eq => continue,
+                .lt => return .lt,
+                .gt => return .gt,
+            }
         }
     }
     return std.math.order(lhs.len, rhs.len);
diff --git a/lib/std/mem.zig b/lib/std/mem.zig
index 6a5e654fab2b..28734d23a681 100644
--- a/lib/std/mem.zig
+++ b/lib/std/mem.zig
@@ -627,12 +627,14 @@ pub fn sortUnstableContext(a: usize, b: usize, context: anytype) void {
 
 /// Compares two slices of numbers lexicographically. O(n).
 pub fn order(comptime T: type, lhs: []const T, rhs: []const T) math.Order {
-    const n = @min(lhs.len, rhs.len);
-    for (lhs[0..n], rhs[0..n]) |lhs_elem, rhs_elem| {
-        switch (math.order(lhs_elem, rhs_elem)) {
-            .eq => continue,
-            .lt => return .lt,
-            .gt => return .gt,
+    if (lhs.ptr != rhs.ptr) {
+        const n = @min(lhs.len, rhs.len);
+        for (lhs[0..n], rhs[0..n]) |lhs_elem, rhs_elem| {
+            switch (math.order(lhs_elem, rhs_elem)) {
+                .eq => continue,
+                .lt => return .lt,
+                .gt => return .gt,
+            }
         }
     }
     return math.order(lhs.len, rhs.len);
@@ -640,6 +642,7 @@ pub fn order(comptime T: type, lhs: []const T, rhs: []const T) math.Order {
 
 /// Compares two many-item pointers with NUL-termination lexicographically.
 pub fn orderZ(comptime T: type, lhs: [*:0]const T, rhs: [*:0]const T) math.Order {
+    if (lhs == rhs) return .eq;
     var i: usize = 0;
     while (lhs[i] == rhs[i] and lhs[i] != 0) : (i += 1) {}
     return math.order(lhs[i], rhs[i]);
@@ -651,6 +654,10 @@ test order {
     try testing.expect(order(u8, "abc", "abc0") == .lt);
     try testing.expect(order(u8, "", "") == .eq);
     try testing.expect(order(u8, "", "a") == .lt);
+
+    const s: []const u8 = "abc";
+    try testing.expect(order(u8, s, s) == .eq);
+    try testing.expect(order(u8, s[0..2], s) == .lt);
 }
 
 test orderZ {
@@ -659,6 +666,9 @@ test orderZ {
     try testing.expect(orderZ(u8, "abc", "abc0") == .lt);
     try testing.expect(orderZ(u8, "", "") == .eq);
     try testing.expect(orderZ(u8, "", "a") == .lt);
+
+    const s: [*:0]const u8 = "abc";
+    try testing.expect(orderZ(u8, s, s) == .eq);
 }
 
 /// Returns true if lhs < rhs, false otherwise
-- 
2.52.0

